eth/beacon/state_machines/base.py

```diff
@@ -193,9 +222,562 @@ def active_state(self) -> ActiveState:
+        """
+        block_committees_info = get_block_committees_info(
+            block,
+            post_crystallized_state,
+            cycle_length,
+        )
+        # Vote
+        attester_bitfield = set_voted(
+            get_empty_bitfield(block_committees_info.proposer_committee_size),
+            block_committees_info.proposer_index_in_committee,
+        )
+
+        # Get justified_slot and justified_block_hash
+        justified_slot = post_crystallized_state.last_justified_slot
+        justified_block_hash = chaindb.get_canonical_block_by_slot(justified_slot).hash
+
+        # Get signing message and sign it
+        parent_hashes = get_hashes_to_sign(
+            post_active_state.recent_block_hashes,
+            block,
+            cycle_length,
+        )
+        message = self.create_signing_message(
+            block.slot_number,
+            parent_hashes,
+            shard_id,
+            shard_block_hash,
+            justified_slot,
+        )
+        sigs = [
+            bls.sign(
+                message,
+                private_key,
+            )
+        ]
+        aggregate_sig = bls.aggregate_sigs(sigs)
+
+        return self.get_attestation_record_class()(
+            slot=block.slot_number,
+            shard_id=shard_id,
+            oblique_parent_hashes=(),
+            shard_block_hash=shard_block_hash,
+            attester_bitfield=attester_bitfield,
+            justified_slot=justified_slot,
+            justified_block_hash=justified_block_hash,
+            aggregate_sig=aggregate_sig,
+        )
+
+    @classmethod
+    def aggregate_attestation_record(cls,
+                                     crystallized_state: CrystallizedState,
+                                     active_state: ActiveState,
+                                     block: BaseBeaconBlock,
+                                     votes: Iterable[Tuple[int, bytes, int]],
+                                     proposer_attestation: 'AttestationRecord',
+                                     cycle_length: int) -> 'AttestationRecord':
+        """
+        Aggregate the votes.
+
+        TODO: Write tests
+        """
+        # Get signing message
+        parent_hashes = get_hashes_to_sign(
+            active_state.recent_block_hashes,
+            block,
+            cycle_length,
+        )
+        message = cls.create_signing_message(
+            block.slot_number,
+            parent_hashes,
+            proposer_attestation.shard_id,
+            block.shard_block_hash,
+            crystallized_state.last_justified_slot,
+        )
+        # Update the bitfield and append the signatures
+        bitfield = proposer_attestation.bitfield
+        sigs = []
+        for (committee_index, sig, public_key) in votes:
+            if bls.verify(message, public_key, sig):
+                bitfield = set_voted(bitfield, committee_index)
+                sigs.append(sig)
```

Comment:
Just looking at what this looks like without the mutation

```python
sigs_with_committe_info = tuple(
    sig, committee_index
    for (committee_index, sig, public_key)
    in votes
    if bls.verify(message, public_key, sig)
)
sigs, commitee_indices = zip(*sigs_with_committe_info)   # may need to handle edge case of empty list
bitfield = pipe(
    proposer_attestation.bitfield
    *(functools.partial(set_voted, index=committee_index) for committee_index in committee_indices)
)
```

That's all psuedo code, probably some minor errors, but it removes the mutation and seems readable enough).  The `functools.partial` can be removed if `set_voted` was made curriable.